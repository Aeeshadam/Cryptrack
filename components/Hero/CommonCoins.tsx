import Box from "@mui/material/Box";
import Grid from "@mui/material/Grid";
import Image from "next/image";
import styled from "@mui/material/styles/styled";

const StyledCoinBox = styled(Box)(({ theme }) => ({
  display: "none",

  [theme.breakpoints.up("md")]: {
    display: "block",
    marginTop: "2rem",
    marginBottom: "2rem",
  },
}));

const CommonCoins = () => {
  return (
    <StyledCoinBox>
      <Grid container spacing={12}>
        <Grid item xs>
          <Image src="/Bitcoin.svg" alt="Bitcoin Icon" width={80} height={80} />
        </Grid>
        <Grid item xs>
          <Image src="/Bnb.svg" alt="BNB Icon" width={80} height={80} />
        </Grid>
        <Grid item xs>
          <Image
            src="/Ethereum.svg"
            alt="Ethereum Icon"
            width={80}
            height={80}
          />
        </Grid>
        <Grid item xs>
          <Image src="/Tether.svg" alt="Tether Icon" width={80} height={80} />
        </Grid>
      </Grid>
    </StyledCoinBox>
  );
};
export default CommonCoins;
