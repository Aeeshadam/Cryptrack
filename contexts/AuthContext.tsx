"use client";
import React, { useEffect } from "react";
import {
  createUserWithEmailAndPassword,
  signInWithEmailAndPassword,
  signOut,
  onAuthStateChanged,
  User,
  updateProfile,
} from "firebase/auth";
import { doc, setDoc, getDoc } from "firebase/firestore";
import db, { auth } from "@/firebase/firebase";
import { createContext, useContext, useState, ReactNode } from "react";
import LoadingSpinner from "@/components/LoadingSpinner";

interface AuthContextProps {
  user: User | null;

  signUp: (email: string, password: string, name: string) => Promise<boolean>;
  signIn: (email: string, password: string) => Promise<boolean>;
  loading: boolean;
  logOut: () => Promise<boolean>;
  displayName: string;
  setError: (error: string) => void;
  // setDisplayName: (name: string | null) => void;
  error: string;
}
const AuthContext = createContext<AuthContextProps | undefined>(undefined);

export const AuthProvider = ({ children }: { children: ReactNode }) => {
  const [user, setUser] = useState<User | null>(null);
  const [loading, setLoading] = useState(true);
  const [displayName, setDisplayName] = useState<string>("");
  const [error, setError] = useState<string>("");

  useEffect(() => {
    const unsubscribe = onAuthStateChanged(auth, async (user) => {
      if (user) {
        setUser(user);
        const docRef = doc(db, "users", user.uid);
        const docSnap = await getDoc(docRef);
        if (docSnap.exists()) {
          setDisplayName(docSnap.data().name);
        }
      } else {
        setUser(null);
        setDisplayName("");
      }
      setLoading(false);
    });
    return unsubscribe;
  }, []);

  const signUp = async (email: string, password: string, name: string) => {
    setLoading(true);
    setError("");
    try {
      const userCredential = await createUserWithEmailAndPassword(
        auth,
        email,
        password
      );
      const user = userCredential.user;
      await updateProfile(user, { displayName: name });
      await setDoc(doc(db, "users", user.uid), {
        name,
        email,
      });
      setDisplayName(name);
      setUser(user);
      return true;
    } catch (error) {
      setError((error as Error).message);
      return false;
    } finally {
      setLoading(false);
    }
  };

  const signIn = async (email: string, password: string) => {
    setLoading(true);
    setError("");
    try {
      const userCredential = await signInWithEmailAndPassword(
        auth,
        email,
        password
      );
      const user = userCredential.user;

      const docRef = doc(db, "users", user.uid);
      const docSnap = await getDoc(docRef);

      if (docSnap.exists()) {
        setDisplayName(docSnap.data().name);
      }
      setUser(user);
      return true;
    } catch (error) {
      setError((error as Error).message);
      return false;
    } finally {
      setLoading(false);
    }
  };

  const logOut = async () => {
    setError("");
    try {
      await signOut(auth);
      setUser(null);
      setDisplayName("");
      return true;
    } catch (error) {
      setError((error as Error).message);
      return false;
    } finally {
    }
  };

  return (
    <AuthContext.Provider
      value={{
        user,
        logOut,
        setError,
        loading,
        error,
        signUp,
        signIn,
        displayName,
      }}
    >
      {children}
    </AuthContext.Provider>
  );
};

export const useAuth = () => {
  const context = useContext(AuthContext);
  if (context === undefined) {
    throw new Error("useAuth must be used within an AuthProvider");
  }
  return context;
};
