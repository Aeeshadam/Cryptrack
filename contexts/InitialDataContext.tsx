"use client";

import {
  createContext,
  useContext,
  useState,
  useEffect,
  ReactNode,
} from "react";
import { InitialDataContextProps, CoinProps } from "@/types";
import { useDispatch } from "react-redux";
import { AppDispatch } from "@/store/store";

const InitialDataContext = createContext<InitialDataContextProps | undefined>(
  undefined
);

export const InitialDataProvider = ({ children }: { children: ReactNode }) => {
  const [initialData, setInitialData] = useState([] as CoinProps[]);
  const dispatch = useDispatch<AppDispatch>();
  useEffect(() => {
    const fetchData = async () => {
      try {
        const response = await fetch(
          `${process.env.NEXT_PUBLIC_BASE_URL}/api/coins`
        );
        if (!response.ok) {
          throw new Error("Failed to fetch coins");
        }
        const data: CoinProps[] = await response.json();
        setInitialData(data);
      } catch (error) {
        console.error("Failed to fetch initial data:", error);
      }
    };

    // fetchData();
  }, [dispatch]);

  return (
    <InitialDataContext.Provider value={{ initialData, setInitialData }}>
      {children}
    </InitialDataContext.Provider>
  );
};

export const useInitialData = () => {
  const context = useContext(InitialDataContext);

  if (!context) {
    throw new Error("useInitialData must be used within a InitialDataProvider");
  }

  return context;
};

export default InitialDataContext;
