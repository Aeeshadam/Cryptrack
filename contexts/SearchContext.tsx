"use client";

import { createContext, useContext, useState, ReactNode } from "react";
import { SearchContextProps } from "@/types";
import { useSelector } from "react-redux";
import { AppState } from "@/store/store";

const SearchContext = createContext<SearchContextProps | undefined>(undefined);

export const SearchProvider = ({ children }: { children: ReactNode }) => {
  const [openSearch, setOpenSearch] = useState(false);
  const [searchQuery, setSearchQuery] = useState("");
  const coins = useSelector((state: AppState) => state.crypto.coins);

  const filteredCoins = searchQuery
    ? coins?.filter((coin) =>
        coin.name.toLowerCase().includes(searchQuery.toLowerCase())
      )
    : coins?.slice(0, 10);

  const handleOpenSearch = () => {
    setOpenSearch(true);
  };
  return (
    <SearchContext.Provider
      value={{
        searchQuery,
        setSearchQuery,
        filteredCoins,
        openSearch,
        setOpenSearch,
        handleOpenSearch,
      }}
    >
      {children}
    </SearchContext.Provider>
  );
};

export const useSearch = () => {
  const context = useContext(SearchContext);
  if (context === undefined) {
    throw new Error("useSearch must be used within a SearchProvider");
  }
  return context;
};
