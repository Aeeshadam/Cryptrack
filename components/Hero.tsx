"use client";
import React from "react";
import Image from "next/image";
import { useRouter } from "next/navigation";


import { StyledButton } from "@/components/StyledButton";
import { useAuth } from "@/contexts/AuthContext";
import { Typography } from "@mui/material";

import {
  HeroBox,
  HeroContainer,
  HeroContent,
  HeroImageContainer,
  HeroSpan,
} from "../styles/HeroStyles";
import CommonCoins from "./CommonCoins";

const Hero = () => {
  const router = useRouter();
  const { user } = useAuth();
  const handleCreatePortfolio = () => {
    if (!user) {
      router.push("/sign-in");
      return;
    }
    router.push("/portfolio");
  };

  return (
    <HeroContainer>
      <HeroBox>
        <HeroContent>
          <Typography variant="h1" marginBottom="1rem">
            The Ultimate
            <HeroSpan> Crypto Portfolio </HeroSpan> Tracker
          </Typography>
          <StyledButton variant="contained" onClick={handleCreatePortfolio}>
            {user ? "View Portfolio" : "Create Portfolio"}
          </StyledButton>
        </HeroContent>
        <HeroImageContainer>
          <Image
            src="/hero.png"
            alt="hero"
            width={400}
            height={400}
            style={{
              width: "100%",
              height: "auto",
              objectFit: "contain",
            }}
          />
        </HeroImageContainer>
      </HeroBox>
      <CommonCoins />
    </HeroContainer>
  );
};
export default Hero;
