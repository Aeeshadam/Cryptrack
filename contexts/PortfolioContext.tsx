"use client";
import React, { createContext, useContext, useState, useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { addCoin, removeCoin } from "../slice/portfolioSlice";
import {
  getPortfolioCoins,
  addCoinToPortfolio,
  updateCoinInPortfolio,
  removeCoinFromPortfolio,
} from "@/firebase/firestoreService";
import { PortfolioCoin, PortfolioContextProps, Transaction } from "@/types";
import { AppDispatch, AppState } from "@/store/store";
import { useAuth } from "@/contexts/AuthContext";
import { useTransaction } from "@/contexts/TransactionContext";

const PortfolioContext = createContext<PortfolioContextProps | undefined>(
  undefined
);

export const PortfolioProvider: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  const { user } = useAuth();
  const [loading, setLoading] = useState(true);
  const dispatch = useDispatch<AppDispatch>();
  const portfolioCoins = useSelector(
    (state: AppState) => state.portfolio.coins
  );
  const coins = useSelector((state: AppState) => state.coinList.coins);

  const {
    selectedCoin,
    setSelectedCoin,
    quantity,
    setQuantity,
    setPricePerCoin,
    pricePerCoin,
    transactionType,
    setOpenModal,
  } = useTransaction();

  const fetchPortfolioCoins = async (userUid: string) => {
    if (!user) return;
    try {
      const fetchedCoins = await getPortfolioCoins(user.uid);
      const validCoins: PortfolioCoin[] = [];

      await Promise.all(
        fetchedCoins.map(async (coin: PortfolioCoin) => {
          if (coin.quantity === 0) {
            await removeCoinFromPortfolio(user.uid, coin.id);
          } else {
            validCoins.push(coin);
            dispatch(addCoin(coin));
          }
        })
      );

      return validCoins;
    } catch (error) {
      console.error("Error fetching and cleaning portfolio coins:", error);
      return [];
    }
  };

  const calculateTotalBalance = () => {
    return portfolioCoins.reduce((acc, portfolioCoin) => {
      const coin: any = coins.find((c) => c.id === portfolioCoin.id);
      if (coin) {
        return acc + portfolioCoin.quantity * coin.current_price;
      }
      return acc;
    }, 0);
  };

  const calculate24HourChangePercentage = () => {
    const { totalValue, totalPreviousValue } = portfolioCoins.reduce(
      (totals, portfolioCoin) => {
        const coin = coins.find((c) => c.id === portfolioCoin.id);
        if (coin) {
          const currentCoinValue = portfolioCoin.quantity * coin.current_price;
          const previousCoinValue =
            portfolioCoin.quantity *
            (coin.current_price / (1 + coin.price_change_percentage_24h / 100));

          totals.totalValue += currentCoinValue;
          totals.totalPreviousValue += previousCoinValue;
        }
        return totals;
      },
      { totalValue: 0, totalPreviousValue: 0 }
    );

    if (totalPreviousValue === 0) return 0;

    const percentageChange =
      ((totalValue - totalPreviousValue) / totalPreviousValue) * 100;
    return percentageChange;
  };

  const handleAddTransaction = (event: React.FormEvent) => {
    event.preventDefault();
    if (!user) return;
    if (selectedCoin && quantity > 0) {
      const coin = coins.find((coin) => coin.id === selectedCoin);
      if (coin) {
        const existingPortfolioCoin = portfolioCoins.find(
          (c) => c.id === coin.id
        );

        const newTransaction: Transaction = {
          quantity,
          price: pricePerCoin,
          type: transactionType,
          timestamp: Date.now(),
        };

        const updatedQuantity =
          transactionType === "buy"
            ? (existingPortfolioCoin?.quantity || 0) + quantity
            : (existingPortfolioCoin?.quantity || 0) - quantity;

        const updatedCoin: PortfolioCoin = {
          id: coin.id,
          name: coin.name,
          quantity: updatedQuantity,
          transactions: existingPortfolioCoin
            ? [...existingPortfolioCoin.transactions, newTransaction]
            : [newTransaction],
        };
        dispatch(addCoin(updatedCoin));

        if (existingPortfolioCoin) {
          updateCoinInPortfolio(user.uid, updatedCoin);
        } else {
          addCoinToPortfolio(user.uid, updatedCoin);
        }
      }
    }

    setOpenModal(false);
    setSelectedCoin("");
    setQuantity(0);
    setPricePerCoin(0);
  };

  const handleRemoveCoin = (coinId: string) => {
    if (!user) return;
    const confirm = window.confirm("Are you sure you want to remove this coin from your portfolio?");
    if (confirm) {
      try {
        removeCoinFromPortfolio(user.uid, coinId);
        dispatch(removeCoin(coinId));
      } catch (error) {
        console.error("Error removing coin from portfolio:", error);
      }
    } else {
      return;
    }
  };

  useEffect(() => {
    const fetchCoins = async () => {
      if (user) {
        try {
          await fetchPortfolioCoins(user.uid);
        } catch (error) {
          console.error("Error fetching portfolio coins: ", error);
        } finally {
          setLoading(false);
        }
      }
    };
    fetchCoins();
  }, [user]);

  return (
    <PortfolioContext.Provider
      value={{
        portfolioCoins,
        coins,
        loading,
        fetchPortfolioCoins,
        handleAddTransaction,
        handleRemoveCoin,
        calculateTotalBalance,
        calculate24HourChangePercentage,
      }}
    >
      {children}
    </PortfolioContext.Provider>
  );
};

export const usePortfolio = () => {
  const context = useContext(PortfolioContext);
  if (!context) {
    throw new Error("usePortfolio must be used within a PortfolioProvider");
  }
  return context;
};
