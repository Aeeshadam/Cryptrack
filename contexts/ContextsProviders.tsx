import ClientProvider from "./StoreProvider";
import { AppRouterCacheProvider } from "@mui/material-nextjs/v13-appRouter";
import { ThemeProvider } from "@mui/material/styles";
import theme from "@/styles/theme";
import CssBaseline from "@mui/material/CssBaseline";
import { SearchProvider } from "./SearchContext";
import { InitialDataProvider } from "./InitialDataContext";

export default function ContextsProviders({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) {
  return (
    <html lang="en">
      <body>
        <AppRouterCacheProvider>
          <ThemeProvider theme={theme}>
            <CssBaseline />
            <ClientProvider>
              <SearchProvider>
                <InitialDataProvider>{children}</InitialDataProvider>
              </SearchProvider>
            </ClientProvider>
          </ThemeProvider>
        </AppRouterCacheProvider>
      </body>
    </html>
  );
}
