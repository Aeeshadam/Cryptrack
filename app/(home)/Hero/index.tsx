"use client";
import React from "react";
import Image from "next/image";
import {
  HeroContainer,
  HeroContent,
  HeroSpan,
  HeroBox,
  HeroImageContainer,
} from "./styles";
import { Typography } from "@mui/material";
import { StyledButton } from "@/components/StyledButton";
import CommonCoins from "./CommonCoins";

const Hero = () => {
  return (
    <HeroContainer>
      <HeroBox>
        <HeroContent>
          <Typography variant="h1" marginBottom="1rem">
            The Ultimate
            <HeroSpan> Cryto Portfolio </HeroSpan> Tracker
          </Typography>
          <StyledButton variant="contained">Create Portfolio</StyledButton>
        </HeroContent>
        <HeroImageContainer>
          <Image
            src="/hero.png"
            alt="hero"
            width={400}
            height={400}
            style={{
              width: "100%",
              height: "auto",
              objectFit: "contain",
            }}
          />
        </HeroImageContainer>
      </HeroBox>
      <CommonCoins />
    </HeroContainer>
  );
};
export default Hero;
