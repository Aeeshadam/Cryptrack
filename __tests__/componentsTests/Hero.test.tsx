import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import Hero from "../../components/Hero";
import { useAuth } from "@/contexts/AuthContext";
import { useRouter } from "next/navigation";

jest.mock("next/navigation", () => ({
  useRouter: jest.fn(),
}));

jest.mock("@/contexts/AuthContext", () => ({
  useAuth: jest.fn(),
}));

describe("Hero", () => {
  const mockPush = jest.fn();

  beforeEach(() => {
    (useRouter as jest.Mock).mockReturnValue({
      push: mockPush,
    });
  });

  it("should render correctly", () => {
    (useAuth as jest.Mock).mockReturnValue({ user: null });
    const { container } = render(<Hero />);
    expect(container).toMatchSnapshot();
  });

  it("navigate to portfolio if user is logged in and button is clicked", () => {
    (useAuth as jest.Mock).mockReturnValue({ user: { uid: "123" } });
    render(<Hero />);
    const button = screen.getByRole("button", { name: /view portfolio/i });
    fireEvent.click(button);
    expect(mockPush).toHaveBeenCalledWith("/portfolio");
  });

  it("navigate to login if user is not logged in and button is clicked", () => {
    (useAuth as jest.Mock).mockReturnValue({ user: null });
    render(<Hero />);
    const button = screen.getByRole("button", { name: /create portfolio/i });
    fireEvent.click(button);
    expect(mockPush).toHaveBeenCalledWith("/sign-in");
  });
});
