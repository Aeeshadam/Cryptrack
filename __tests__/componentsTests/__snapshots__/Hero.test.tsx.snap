// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hero should render correctly 1`] = `
<div>
  <section
    class="css-1ptp9m9"
  >
    <div
      class="css-1vyekul"
    >
      <div
        class="css-1ibemvj"
      >
        <h1
          class="MuiTypography-root MuiTypography-h1 css-3h6rmx-MuiTypography-root"
        >
          The Ultimate
          <span
            class="css-1bm9945"
          >
             Crypto Portfolio 
          </span>
           Tracker
        </h1>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-1944k86-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Create Portfolio
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="css-frdxje"
      >
        <img
          alt="hero"
          data-nimg="1"
          decoding="async"
          height="400"
          loading="lazy"
          src="/_next/image?url=%2Fhero.png&w=828&q=75"
          srcset="/_next/image?url=%2Fhero.png&w=640&q=75 1x, /_next/image?url=%2Fhero.png&w=828&q=75 2x"
          style="color: transparent; width: 100%; height: auto; object-fit: contain;"
          width="400"
        />
      </div>
    </div>
    <div
      class="MuiBox-root css-jhoeoz"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-12 css-1fh410u-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
        >
          <img
            alt="Bitcoin Icon"
            data-nimg="1"
            decoding="async"
            height="80"
            loading="lazy"
            src="/Bitcoin.svg"
            style="color: transparent;"
            width="80"
          />
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
        >
          <img
            alt="BNB Icon"
            data-nimg="1"
            decoding="async"
            height="80"
            loading="lazy"
            src="/Bnb.svg"
            style="color: transparent;"
            width="80"
          />
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
        >
          <img
            alt="Ethereum Icon"
            data-nimg="1"
            decoding="async"
            height="80"
            loading="lazy"
            src="/Ethereum.svg"
            style="color: transparent;"
            width="80"
          />
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
        >
          <img
            alt="Tether Icon"
            data-nimg="1"
            decoding="async"
            height="80"
            loading="lazy"
            src="/Tether.svg"
            style="color: transparent;"
            width="80"
          />
        </div>
      </div>
    </div>
  </section>
</div>
`;
